### PR TITLE
EQ 622: Household Composition Summary

### LINKED ISSUE #629: Census: For Question 3 When respondent select 'No - I need to add another person' gives an error 500
### Expected behaviour

No - return to question 2 to review updated household members and add another person. 

### Actual behaviour
Error 500

### Steps to reproduce the behaviour

A branch is created with a selenium test to reproduce the issue.

https://github.com/ONSdigital/eq-survey-runner/tree/respone-no-need-to-add-another-person

yarn run test_functional -- --spec=census/issues/respone-no-need-to-add-another-person

#### Browser
All

### PR DESCRIPTION
**Pull request by iwootten**
### What is the context of this PR?
Adds the household composition summary page, showing a list of users entered and giving the option to return to edit the list.

In order to support this I've added conditional routing to a previous location. Allows us to support the routing of users to the household composition screen in census, based on the answer to a question. This is handled by adding a route to the routing path and removing the users current answer to this question (which would otherwise prevent the location path being calculated by navigator). This also fixes #629.

### How to review 
There are unit tests in navigator and integration tests to cover new behaviour. I've had to update current composition selenium tests to work with the new block.

### CODE REVIEW

File: app/data/test_household_question.json [hunk 1]
```diff
@@ -61,6 +62,95 @@
...
+                    },
+                    "sections": [
+                        {
+                            "display": {
+                                "properties": {}
+                            },
+                            "id": "further-section",
+                            "questions": [
+                                {
+                                  "description": "",
```
Comment (benjefferies) [inline app/data/test_household_question.json]: Can description/display be removed?

File: app/data/test_household_question.json [hunk 2]
```diff
@@ -61,6 +62,95 @@
...
                     "title": "Household",
                     "validation": []
+                },
+                {
+                    "display": {
+                        "properties": {}
+                    },
+                    "sections": [
+                        {
+                            "display": {
```
Comment (benjefferies) [inline app/data/test_household_question.json]: Can display be removed?

File: app/jinja_filters.py [hunk 3]
```diff
@@ -49,3 +49,10 @@ def format_str_as_month_year_date(value):
 @blueprint.app_template_filter()
 def format_str_as_date(value):
     return format_str_as_short_date(value)
+
+
+@blueprint.app_template_filter('pluralize')
+def pluralize(number, singular='', plural='s'):
```
Comment (benjefferies) [inline app/jinja_filters.py]: Should we not be able to get this functionality from flask babel?

File: app/main/views/questionnaire.py [hunk 4]
```diff
@@ -251,26 +246,9 @@ def post_household_composition(eq_id, form_type, collection_id, group_id):
     if not valid:
         return _render_template(questionnaire_manager.state, group_id, 0, 'household-composition', template='questionnaire')
 
-    return _go_to_next_block(location=this_block, eq_id=eq_id,
-                             form_type=form_type, collection_id=collection_id)
-
+    next_location = navigator.get_next_location(current_block_id='household-composition', current_iteration=0, current_group_id=group_id)
 
-def _go_to_next_block(location, eq_id, form_type, collection_id):
```
Comment (benjefferies) [inline app/main/views/questionnaire.py:264]: Nice tidy up

File: app/questionnaire/navigator.py [hunk 5]
```diff
@@ -104,54 +19,74 @@ def __init__(self, survey_json, metadata=None, answer_store=None):
...
         self.first_group_id = SchemaHelper.get_first_group_id(self.survey_json)
         self.last_group_id = SchemaHelper.get_last_group_id(self.survey_json)
+
+        self.first_block_location = {
+            "group_id": self.first_group_id,
+            "group_instance": 0,
+            "block_id": self.first_block_id,
+        }
         self.location_path = self.get_location_path()
+        self.first_location = self.location_path[0]
```
Comment (benjefferies) [inline app/questionnaire/navigator.py:33]: Aren't first_block_location and first_location the same?

File: app/questionnaire/navigator.py [hunk 6]
```diff
@@ -104,54 +19,74 @@ def __init__(self, survey_json, metadata=None, answer_store=None):
...
-        # If this isn't the last block in the set evaluated
-        elif block_id_index != len(blocks) - 1:
-            next_block_id = blocks[block_id_index + 1]['block']['id']
-            next_group_id = blocks[block_id_index + 1]['group_id']
-            next_group_instance = blocks[block_id_index + 1]['group_instance']
+                    next_block_index = self._block_index_for_location(blocks, next_location)
+
+                    # We're jumping backwards, so need to delete current answer
+                    if next_block_index is not None and block_index > next_block_index:
+                        answer = Answer(answer_id=rule['goto']['when']['id'],
```
Comment (benjefferies) [inline app/questionnaire/navigator.py:77]: Do we need another check here incase it's a meta rule?

File: app/data/test_household_question.json [hunk 7]
```diff
@@ -61,6 +62,95 @@
...
+                                            "validation": {
+                                                "messages": {}
+                                            }
+                                        }
+                                  ]
+
+                                }
+                            ],
+                            "title": "Who lives here?",
+                            "description": "<h3>There {{answers.person_full_name|length|pluralize('is', 'are')}} <strong class='strong-census'>{{answers.person_full_name|length}} {{answers.person_full_name|length|pluralize('person', 'people')}}</strong> in your household. </h3><div class='name-of-person'>{% for name in answers.person_full_name %}<div id='entity-list__{{loop.index}}'><div class='box--entity__person mars'>{{name}}</div></div>{% endfor %}</div>",
```
Comment (jonnyshaw89) [inline app/data/test_household_question.json]: We need to also integrate this into the census household schema.

File: tests/functional/pages/surveys/household_composition/summary.page.js [hunk 8]
```diff
@@ -2,18 +2,23 @@ import SummaryPage from '../../summary.page'
...
+  }
+
+  submit() {
+    browser.click('.qa-btn-submit')
     return this
   }
 
   isNameDisplayed(name) {
-    return browser.element('[data-qa="household-full-name-answer"]').getText() === name
+    return browser.element('.box--entity__person').getText() === name
```
Comment (benjefferies) [inline tests/functional/pages/surveys/household_composition/summary.page.js]: I think the preference to to use id/name/attribute hook over classes

File: app/main/views/questionnaire.py [hunk 9]
```diff
@@ -143,10 +143,7 @@ def post_interstitial(eq_id, form_type, collection_id, block_id):
 
     logger.info("Redirecting user to next location %s with tx_id=%s", str(next_location), metadata["tx_id"])
 
-    return redirect(block_url(eq_id, form_type, collection_id,
-                              group_id=next_location['group_id'],
-                              group_instance=next_location['group_instance'],
-                              block_id=next_location['block_id']))
+    return redirect(location_url(eq_id, form_type, collection_id, next_location))
```
Comment (jonnyshaw89) [inline app/main/views/questionnaire.py:147]: 👍

**Review by benjefferies — CHANGES_REQUESTED** [review]

**Review by jonnyshaw89 — CHANGES_REQUESTED** [review]

File: app/questionnaire/navigator.py (hunk 5 above)
Comment (iwootten) [reply app/questionnaire/navigator.py:33]: Nope, not always. First_location points to everything, including interstital. Block location only includes blocks.

File: tests/functional/pages/surveys/household_composition/summary.page.js (hunk 8 above)
Comment (iwootten) [reply tests/functional/pages/surveys/household_composition/summary.page.js]: Yeah, this is subject to however we decide the template we're rendering is going to end up.

File: app/questionnaire/navigator.py [hunk 10]
```diff
@@ -104,54 +19,74 @@ def __init__(self, survey_json, metadata=None, answer_store=None):
         self.first_block_id = SchemaHelper.get_first_block_id(self.survey_json)
```
Comment (benjefferies) [inline app/questionnaire/navigator.py]: We don't use this first_block_id other than below to create the first_block_location, maybe we should get rid as we have lots of first location related fields

**Comment by benjefferies** [comment]
Can you cherry pick issue selenium test onto branch https://github.com/ONSdigital/eq-survey-runner/commit/ecabc7f571970d17d5ebaf0f4bc203c8325f6925

**Comment by iwootten** [comment]
@benjefferies - the scenarios are covered in the current census/routing/census-routing-scenarios.spec.js routing file and commented out, I've just uncommented everything from there.

**Comment by iwootten** [comment]
I've reverted the changes I made to repeating answers and left with guidance displaying incorrectly. 

screen shot 2016-12-09 at 10 16 05

File: tests/functional/spec/census/routing/census-routing-scenarios.spec.js [hunk 11]
```diff
@@ -76,516 +76,515 @@
...
-    it('Given I am answering question 3 in the who lives here section, When I dont select any respone, Then I am routed to Who lives here question 4 ', function () {
+    it('Given I am answering question 3 in the who lives here section, When I dont select any response, Then I am routed to Who lives here question 4 ', function () {
         startCensusQuestionnaire('census_household.json', true)
         PermanentOrFamilyHome.clickPermanentOrFamilyHomeAnswerYes().submit()
         HouseholdComposition.setPersonName(0, 'John Smith').addPerson().setPersonName(1, 'Jane Smith').submit()
         EveryoneAtAddressConfirmation.submit()
         OvernightVisitors.setOvernightVisitorsAnswer(0).submit()
-    }) */
-
-    it('Given I am answering question 1 in the individual detail section, When I select -yes- as respone, Then I am routed to What is you sex question 2 ', function () {
```
Comment (benjefferies) [inline tests/functional/spec/census/routing/census-routing-scenarios.spec.js]: How come we are deleting these tests?

**Review by jonnyshaw89 — APPROVED** [review]

**Review by benjefferies — APPROVED** [review]

**Comment by benjefferies** [comment]
Has anyone front-end reviewed? @hamishtaplin @bennpearson

**Comment by benjefferies** [comment]
If someone okays the last commit I'll squash down and someone can merge

File: app/data/test_household_question.json [hunk 12]
```diff
@@ -82,11 +81,87 @@
...
+                                    "when": {
+                                        "id" : "household-composition-add-another",
+                                        "condition": "equals",
+                                        "value":"No"
+                                    }
+                                }
+                            },
+                            {
+                                "goto": {
+                                    "id": "summary",
```
Comment (collisdigital) [inline app/data/test_household_question.json:151]: Is this right in terms of where to go next?

Comment (benjefferies) [reply app/data/test_household_question.json:151]: In the test schema does it matter much?

Comment (benjefferies) [reply app/data/test_household_question.json:151]: I think it makes sense in the context of the question, if you have entered everyone go to the next section which in the test schema is the summary

Comment (collisdigital) [reply app/data/test_household_question.json:151]: Sorry thought it was the HH schema, ignore me 😆

**Comment by jonnyshaw89** [comment]
👍 LGTM

**Comment by benjefferies** [comment]
Squashed and ready to go once all is green

**Pull request merged**